compiler/qsc_partial_eval/src/lib.rs

```diff
@@ -0,0 +1,710 @@
+        Self {
+            package_store,
+            compute_properties,
+            eval_context: context,
+            assigner,
+            backend: QubitsAndResultsAllocator::default(),
+            callables_map: FxHashMap::default(),
+            program,
+            error: None,
+        }
+    }
+
+    fn create_intrinsic_callable(
+        &self,
+        store_item_id: StoreItemId,
+        callable_decl: &CallableDecl,
+    ) -> Callable {
+        let callable_package = self.package_store.get(store_item_id.package);
+        let name = callable_decl.name.name.to_string();
+        let input_type: Vec<rir::Ty> = callable_package
+            .derive_callable_input_params(callable_decl)
+            .iter()
+            .map(|input_param| map_fir_type_to_rir_type(&input_param.ty))
+            .collect();
+        let output_type = if callable_decl.output == Ty::UNIT {
+            None
+        } else {
+            Some(map_fir_type_to_rir_type(&callable_decl.output))
+        };
+        let body = None;
+        let call_type = if name.eq("__quantum__qis__reset__body") {
+            CallableType::Reset
+        } else {
+            CallableType::Regular
+        };
+        Callable {
+            name,
+            input_type,
+            output_type,
+            body,
+            call_type,
+        }
+    }
+
+    fn eval(mut self) -> Result<Program, Error> {
+        let current_package = self.get_current_package_id();
+        let entry_package = self.package_store.get(current_package);
+        let Some(entry_expr_id) = entry_package.entry else {
+            panic!("package does not have an entry expression");
+        };
+
+        // Visit the entry point expression.
+        self.visit_expr(entry_expr_id);
+
+        // If there was an error, return it.
+        if let Some(error) = self.error {
+            return Err(error);
+        }
+
+        // Insert the return expression and return the generated program.
+        let current_block = self
+            .program
+            .blocks
+            .get_mut(self.eval_context.current_block)
+            .expect("block does not exist");
+        current_block.0.push(Instruction::Return);
+        Ok(self.program)
+    }
+
+    fn eval_callee_expr(
+        &mut self,
+        callee_expr_id: ExprId,
+    ) -> (StoreItemId, FunctorApp, CallableDecl) {
+        let current_package_id = self.get_current_package_id();
+        let store_callee_expr_id = StoreExprId::from((current_package_id, callee_expr_id));
+
+        // Verify that the callee expression is classical.
+        let callable_scope = self.eval_context.get_current_scope();
+        let callee_expr_generator_set = self.compute_properties.get_expr(store_callee_expr_id);
+        let callee_expr_compute_kind = callee_expr_generator_set
+            .generate_application_compute_kind(&callable_scope.args_runtime_properties);
+        assert!(matches!(callee_expr_compute_kind, ComputeKind::Classical));
```

Comment:
nitpick: for this and a couple other asserts in the file, it's worth including a string for the panic message just to help guide investigation; something like `assert!(condition, "compute kind must be classical")`. It's a little redundant for checks as clear as this one, but I think still worth it for future devs (including us!) who might somehow break it.

Reply:
Added messages to asserts for the benefit of our future selves!